falcon/request.py

```diff
@@ -1617,6 +1664,85 @@ def _get_wrapped_wsgi_input(self):
 
         return helpers.BoundedStream(self.env['wsgi.input'], content_length)
 
+    def _getFieldValue(self, field):
+        """Returns input value for imput text fields and FileStream object
+        for input file and None if no file is selected for upload
+
+        Args:
+            field (cgi.FieldStorage or cgi.MiniFieldStorage).
+        """
+        if field.file and 'filename' in field.disposition_options:
+
+            # read first byte of file buffer to make sure if file content exists
+            first_byte = field.file.read(1)
+
+            if first_byte:
+                # NOTE(kgriffs): This is an uploaded file.
+                return FileStream(
+                    file_buffer=field,
+                    max_size=self.options.file_max_upload_size,
+                    first_byte=first_byte
+                )
+            else:
+                # NOTE(kgriffs): No contents found which means no file uploaded.
+                return None
+        else:
+            # NOTE(kgriffs): This is an text field.
+            # DECODE_HTML_CHARS is used to convert html code to utf-8 charcters
+            return DECODE_HTML_CHARS(field.value)
+
+    def _parse_form_urlencoded_or_multipart_form(self):
+        """Parses urlencoded or multipart form data.
+        It sets text fileds in _form_data dict and file buffers in _files dict.
+        """
+
+        if (
+            self.method not in ('GET', 'HEAD') and
+            self.content_type is not None and
+            (
+                'multipart/form-data' in self.content_type or
+                'application/x-www-form-urlencoded' in self.content_type
+            )
+        ):
+            # NOTE(kgriffs): Store input params in _form_data and files in _files
+            self._files = {}
+            self._form_data = {}
+        else:
+            return
+
+        field_storage = cgi.FieldStorage(fp=self.stream, environ=self.env)
```

Comment:
cgi.FieldStorage will blow up if the env doesn't include QUERY_STRING (see also: https://gist.github.com/kgriffs/fe371bfbb2d63211889b#file-gistfile1-py-L9)